### ISSUE TITLE
Add a `GOceanInlineTrans`

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSyclone#2415]
We can't resolve the Field type of a member of r2d_field so the generic inline trans won't inline, but its GOcean so we know e.g. %data is an array so we can check that and resolve the inline trans in that case.

Hopefully this can just be done with a new validate on a new transformation.

**Comment by sergisiso** [comment]
Ideally the other two issues you raised (to give r2d_field info and to make sure its accessible) will already solve the issues with generic inlining?

**Comment by rupertford** [comment]
It would required some GOcean-specific inlining code and would be nicer with a GOcean-specific r2d_field type (as we do for LFRic types in kernels) but could we not recognise the r2d_field derived type and then treat that in a special way when we inline?

**Comment by arporter** [comment]
> It would required some GOcean-specific inlining code and would be nicer with a GOcean-specific r2d_field type (as we do for LFRic types in kernels) but could we not recognise the r2d_field derived type and then treat that in a special way when we inline?

Yes, I think that's what we're proposing. Essentially, if an argument is a GOcean field then we know what its type is...

**Comment by hiker** [comment]
I am closing this as a duplicate of #2705 (and the corresponding PR #3218 is ready for review now).
